src/provider/model/Model.ts

```diff
@@ -277,17 +244,48 @@ export class Model {
         return this.windows.filter(predicate);
     }
 
-    private getOrCreateExpectedWindow(identity: Identity): ExpectedWindow {
+    private getOrCreateExpectedWindow(identity: Identity, windowSeen?: 'seen'): ExpectedWindow {
         const id = getId(identity);
 
         if (this._expectedWindows.has(id)) {
             return this._expectedWindows.get(id)!;
         } else {
-            const expectedWindow: ExpectedWindow = {
-                pendingDeferredPromise: new DeferredPromise<Boxed<Promise<AppWindow>>>(),
-                createdDeferredPromise: new DeferredPromise<void>(),
-                registerDeferredPromise: new DeferredPromise<AppWindow>()
-            };
+            // Create a promise that resolves when the window has registered, or rejects if it is closed
+            const registered = Promise.race([
+                untilTrue((args) => {
+                    return !!this._windowsById[id];
+                }, this._onWindowRegisteredInternal).then(() => {
+                    return this._windowsById[id];
+                }),
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
+                }, this._environment.windowClosed).then(() => {
+                    throw new Error(EXPECT_CLOSED_MESSAGE);
+                })
+            ]);
+
+            // Create a promise that resolves when the window has connected, or rejects if it is closed
+            const connected = Promise.race([
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
+                }, this._apiHandler.onConnection),
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
+                }, this._environment.windowClosed).then(() => {
+                    throw new Error(EXPECT_CLOSED_MESSAGE);
+                })
+            ]);
+
+            // Create a promise that resovles once the window has been seen, to a promise representing successful registration within our timeout
+            const seen = (windowSeen === 'seen' ? Promise.resolve() : untilTrue((args) => {
+                return args !== undefined && getId(args[0]) === id;
+            }, this._environment.windowSeen)).then(() => {
+                return {value: withStrictTimeout(Timeouts.WINDOW_SEEN_TO_REGISTERED, registered, EXPECT_TIMEOUT_MESSAGE).then(() => {
+                    return this._windowsById[id];
```

Comment:
Hard to follow the logic here, maybe split this into 2 statements? 1 creating the promise/then and then the return
```ts
() => {
	const value = withStrictTimeout...
	return {value}
}
```

Reply:
Changed this bit quite a lot. Should be easy to follow now